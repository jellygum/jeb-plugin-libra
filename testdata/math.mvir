module Math {

  public sum_(n: u64, acc: u64): u64 {
    let new_n: u64;
    let new_acc: u64;
    let new_sum: u64;

    if (copy(n) == 0) {
      return move(acc);
    }

    new_n = copy(n) - 1;
    new_acc = move(acc) + move(n);
    new_sum = Self.sum_(move(new_n), move(new_acc));

    return move(new_sum);
  }

  public sum(n: u64): u64 {
    let result: u64;
    result = Self.sum_(move(n), 0);
    return move(result);
  }

}