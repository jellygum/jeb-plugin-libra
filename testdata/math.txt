CompiledModule: {
Module Handles: [
	0x0.Math,]
Struct Handles: []
Function Handles: [
	0x0.Math.sum_(Integer, Integer): (Integer),
	0x0.Math.sum(Integer): (Integer),]
Struct Definitions: []
Field Definitions: []
Function Definitions: [
	public 0x0.Math.sum_(Integer, Integer): (Integer)
		locals(0): Integer, Integer, Integer, Integer, Integer,
		CopyLoc(0)
		LdConst(0)
		Eq
		BrFalse(6)
		MoveLoc(1)
		Ret
		CopyLoc(0)
		LdConst(1)
		Sub
		StLoc(2)
		MoveLoc(1)
		MoveLoc(0)
		Add
		StLoc(3)
		MoveLoc(2)
		MoveLoc(3)
		Call(0x0.Math.sum_(Integer, Integer): (Integer))
		StLoc(4)
		MoveLoc(4)
		Ret,
	public 0x0.Math.sum(Integer): (Integer)
		locals(1): Integer, Integer,
		MoveLoc(0)
		LdConst(0)
		Call(0x0.Math.sum_(Integer, Integer): (Integer))
		StLoc(1)
		MoveLoc(1)
		Ret,]
Type Signatures: []
Function Signatures: [
	(Integer, Integer): (Integer),
	(Integer): (Integer),]
Locals Signatures: [
	Integer, Integer, Integer, Integer, Integer,
	Integer, Integer,]
Strings: [
	Math,
	sum_,
	sum,]
ByteArrays: []
Addresses: [
	0x0,]
}

