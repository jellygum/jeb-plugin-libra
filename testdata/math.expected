Module Handles: [
    0x0.Math,
]
Struct Handles: [
]
Function Handles: [
    0x0.Math.sum_(integer, integer): (integer),
    0x0.Math.sum(integer): (integer),
]
Struct Definitions: [
]
Field Definitions: [
]
Type Signatures: [
]
Function Signatures: [
    (integer, integer): (integer),
    (integer): (integer),
]
Local Signatures: [
    integer, integer, integer, integer, integer,
    integer, integer,
]
Strings: [
    Math,
    sum_,
    sum,
]
ByteArrays: [
]
Addresses: [
    0x0,
]
Function Definitions: [
    public 0x0.Math.sum_(integer, integer): (integer)
        max_stack_size: 3
        locals: integer, integer, integer, integer, integer
        COPY_LOC IdxLocal(0)               [0,1]
        LD_CONST ImmUint64(0)              [1,2]
        EQ                                 [2,1]
        BR_FALSE Branch(6)                 [1,0]
        MOVE_LOC IdxLocal(1)               [0,1]
        RET                                [1,0]
        COPY_LOC IdxLocal(0)               [0,1]
        LD_CONST ImmUint64(1)              [1,2]
        SUB                                [2,1]
        ST_LOC IdxLocal(2)                 [1,0]
        MOVE_LOC IdxLocal(1)               [0,1]
        MOVE_LOC IdxLocal(0)               [1,2]
        ADD                                [2,1]
        ST_LOC IdxLocal(3)                 [1,0]
        MOVE_LOC IdxLocal(2)               [0,1]
        MOVE_LOC IdxLocal(3)               [1,2]
        CALL IdxFuncHandle(0)              [2,1]
        ST_LOC IdxLocal(4)                 [1,0]
        MOVE_LOC IdxLocal(4)               [0,1]
        RET                                [1,0]
    ,
    public 0x0.Math.sum(integer): (integer)
        max_stack_size: 2
        locals: integer, integer
        MOVE_LOC IdxLocal(0)               [0,1]
        LD_CONST ImmUint64(0)              [1,2]
        CALL IdxFuncHandle(0)              [2,1]
        ST_LOC IdxLocal(1)                 [1,0]
        MOVE_LOC IdxLocal(1)               [0,1]
        RET                                [1,0]
    ,
]

